### PR TITLE
Issue 70

### PR DESCRIPTION
**Pull request by baranowb**
New, simpler iteration of PR processor.
https://github.com/jboss-set/pull-request-processor/issues/70

### CODE REVIEW

File: src/main/java/org/jboss/set/pull/processor/ProcessorConfig.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private List<StreamDefinition> writePermitedStreamDefinition;
+    // root dir where we can work.
+    private String rootDirectory;
+
+    private boolean write = false;
+
+    public ProcessorConfig(final List<Evaluator> evaluators, final List<Action> actions, final List<StreamDefinition> streamDefinition,
+            final List<StreamDefinition> writePermitedStreamDefinition, final Aphrodite aphrodite, final ExecutorService executorService,
+            final String rootDirectory, final boolean write) {
+        super();
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/ProcessorConfig.java]: unwanted?

File: src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+/**
+ * Abstract class. Base for processors. Provide basic code to allow simpler processor dev.
+ *
+ * @author baranowb
+ *
+ */
+public abstract class AbstractProcessor implements Processor {
+
+    protected static final Logger logger = Logger.getLogger("org.jboss.pull.processor.processes");
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java]: This is not the package name.

File: src/main/java/org/jboss/set/pull/processor/Main.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+import org.jboss.set.aphrodite.spi.NotFoundException;
+
+import net.sourceforge.argparse4j.ArgumentParsers;
+import net.sourceforge.argparse4j.inf.ArgumentParser;
+import net.sourceforge.argparse4j.inf.ArgumentParserException;
+import net.sourceforge.argparse4j.inf.Namespace;
+
+public class Main {
+
+    public static Logger logger = Logger.getLogger("org.jboss.pull.processor");
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/Main.java]: This is not the package name.

File: src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                // ie. report write will be per processor
+                action.execute(actionContext, processedPullRequests);
+            }
+        } catch (InterruptedException ex) {
+            throw new ProcessorException("processor execution failed", ex);
+        }
+    }
+
+    protected void log(final Level level, final String msg) {
+        this.logger.log(level, this.simpleName + " " + msg);
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java]: warning: The static field AbstractProcessor.logger should be accessed in a static way.

File: src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    protected void log(final Level level, final String msg, final Throwable t) {
+        this.logger.log(level, this.simpleName + " " + msg, t);
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/AbstractProcessor.java]: warning: The static field AbstractProcessor.logger should be accessed in a static way.

File: src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+public class EvaluatorContext {
+
+    private final Aphrodite aphrodite;
+
+    private final PullRequest pullRequest;
+
+    private final StreamComponentDefinition streamComponentDefinition;
+
+    public EvaluatorContext(final Aphrodite aphrodite, final Repository repository, final PullRequest pullrequest, final StreamComponentDefinition streamComponentDefinition) {
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java]: parameter repository is useless.

File: src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    public Aphrodite getAphrodite() {
+        return aphrodite;
+    }
+
+    public PullRequest getPullRequest() {
+        return this.pullRequest;
+    }
+
+    public String getBranch() {
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java]: getBranch() is never called. unwanted?

File: src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedIssuesEvaluator.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            if (upstreamIssue == null && context.getAphrodite().isUpstreamRequired(context.getPullRequest())) {
+                upstreamIssueData.notRequired();
+            }
+            // TODO: XXX what if no upstream required but its there?
+            data.setAttributeValue(EvaluatorData.Attributes.ISSUE_CURRENT, currentIssueData);
+            data.setAttributeValue(EvaluatorData.Attributes.ISSUE_UPSTREAM, upstreamIssueData);
+            data.setAttributeValue(EvaluatorData.Attributes.ISSUES_RELATED, relatedIssiesData);
+        } catch (MalformedURLException | NotFoundException e) {
+            //TODO: XXX remove this in favor of proper reporting
+            e.printStackTrace();
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedIssuesEvaluator.java]: better log than print

File: src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedPullRequestEvaluator.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            final PullRequestData upstreamPullRequestData = convert(upstreamPullRequest, null); // TODO: check if we can fetch
+                                                                                                // StreamDef
+                                                                                                // for this?
+            final IssueData upstreamIssue = data.getAttributeValue(EvaluatorData.Attributes.ISSUE_UPSTREAM);
+            if (!upstreamIssue.isRequired())
+                upstreamPullRequestData.notRequiered();
+            data.setAttributeValue(EvaluatorData.Attributes.PULL_REQUEST_UPSTREAM, upstreamPullRequestData);
+        } catch (MalformedURLException | NotFoundException e) {
+            //TODO: XXX remove this in favor of proper reporting
+            e.printStackTrace();
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedPullRequestEvaluator.java]: better log than print

File: src/main/java/org/jboss/set/pull/processor/ActionContext.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                }
+                try {
+                    if (repoURL.toURI().equals(streamComponentDefinition.getStreamComponent().getRepositoryURL())
+                            && pullRequestBranch
+                                    .equals(streamComponentDefinition.getStreamComponent().getCodebase().getName())) {
+                        return true;
+                    }
+                } catch (URISyntaxException e) {
+                    // TODO: XXX this should not happen, add proper log/info output for CI console
+                    e.printStackTrace();
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/ActionContext.java]: better log than print

File: src/main/java/org/jboss/set/pull/processor/impl/process/OpenPRProcessor.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                            final List<PullRequest> componentPullRequests = super.processorConfig.getAphrodite().getPullRequestsByState(repository, PullRequestState.OPEN);
+                            //translate it into refs, add to ret val
+                            pullRequests.addAll(componentPullRequests.stream().map(p->{
+                                PullRequestReference pullRequestReference = new PullRequestReference();
+                                pullRequestReference.setComponentDefinition(streamComponentDefinition);
+                                pullRequestReference.setPullRequest(p);
+                                return pullRequestReference;
+                            }).collect(Collectors.toList()));
+                        } else {
+                            super.logger.warning("Did not find repository: "+streamComponentDefinition.getStreamComponent().getRepositoryURL());
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/process/OpenPRProcessor.java]: warning: The static field AbstractProcessor.logger should be accessed in a static way

File: src/main/java/org/jboss/set/pull/processor/impl/evaluator/UpstreamPullRequestLabelEvaluator.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    public void eval(EvaluatorContext context, EvaluatorData data) {
+        final PullRequestData upstreamPullRequestData = data.getAttributeValue(EvaluatorData.Attributes.PULL_REQUEST_UPSTREAM);
+        LabelData labelData = super.getLabelData(EvaluatorData.Attributes.LABELS_UPSTREAM, data);
+
+        if (upstreamPullRequestData.isDefined()) {
+            if (upstreamPullRequestData.isMerged()) {
+                labelData.addLabelItem(new LabelItem(LabelItem.LabelContent.Upstream_merged, LabelItem.LabelAction.SET,
+                        LabelItem.LabelSeverity.OK));
+            } else {
+                // TODO: check if this can happen
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/evaluator/UpstreamPullRequestLabelEvaluator.java]: always jump into else block as PullRequestData.isMerged() is unimplemented.

File: src/main/java/org/jboss/set/pull/processor/impl/action/SetLabelsAction.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+    private static final Logger LOG = Logger.getLogger(SetLabelsAction.class.getName());
+
+    @Override
+    public void execute(final ActionContext actionContext, final List<EvaluatorData> data) {
+        try {
+            actionContext.getExecutors().invokeAll(
+                    data.stream().map(e -> new EvaluatorProcessingTask(actionContext, e)).collect(Collectors.toList()));
+        } catch (InterruptedException ex) {
+            ex.printStackTrace();
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/action/SetLabelsAction.java]: better log than print

File: src/main/java/org/jboss/set/pull/processor/impl/action/SetLabelsAction.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            final StringBuilder logBuilder = new StringBuilder();
+            logBuilder.append("\n... " + pullRequestData.getPullRequest().getURL());
+            logBuilder.append("\n   |... " + (issueData.isDefined() ? issueData.getIssue().getURL() : "n/a"));
+            logBuilder.append("\n   |... C:" + currentLabels.stream().map(l -> l.getName()).collect(Collectors.toList()));
+            logBuilder.append("\n   |... S:" + addList.stream().map(l -> l.getLabel()).collect(Collectors.toList()));
+            logBuilder.append("\n   |... R:" + removeList.stream().map(l -> l.getLabel()).collect(Collectors.toList()));
+
+            if (upstreamPullRequestData.isDefined()) {
+                final Set<Label> upstreamLabels = new TreeSet(new LabelComparator());
+                currentLabels.addAll(
```
Comment (soul2zimate) [inline src/main/java/org/jboss/set/pull/processor/impl/action/SetLabelsAction.java]: Should this be upstreamLabels.addAll ?

File: src/main/java/org/jboss/set/pull/processor/data/LabelData.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            throw new IllegalArgumentException(li.toString());
+        }
+        this.labels.add(li);
+    }
+
+    public List<LabelItem> getLabels(LabelItem.LabelAction act) {
+        return labels.stream().filter(l -> l.getAction().equals(act)).collect(Collectors.toList());
+    }
+
+    public Collection<LabelItem> getLabels() {
```
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/data/LabelData.java]: ArrayList is collection, isnt it :) ? But yeah, I guess there is no need to generalize, List should be fine.

File: src/main/java/org/jboss/set/pull/processor/data/PullRequestData.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.set.pull.processor.data;
+
+import org.jboss.set.aphrodite.domain.PullRequest;
+import org.jboss.set.pull.processor.StreamComponentDefinition;
+
+public class PullRequestData {
```
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/data/PullRequestData.java]: No need for it, no benefit other than confusion, I think.

File: src/main/java/org/jboss/set/pull/processor/StreamComponentDefinition.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+    public StreamComponent getStreamComponent() {
+        return streamComponent;
+    }
+
+    public void setStreamComponent(StreamComponent streamComponent) {
+        this.streamComponent = streamComponent;
+    }
+
+    public StreamDefinition getStreamDefinition() {
```
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/StreamComponentDefinition.java]: To fetch stream this component belongs to.

File: src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java (hunk 6 above)
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java]: Yes, missed that on clean up.

File: src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java (hunk 7 above)
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/EvaluatorContext.java]: Nope. We need comparison between branch and component/stream - just not implemented yet.

File: src/main/java/org/jboss/set/pull/processor/Main.java [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+                        .collect(Collectors.toList());
+                final ProcessorConfig processorConfig = new ProcessorConfig(filteredEvaluators, filteredActions, parsedStreams, writePermittedStreams,
+                        aphrodite, executor.executorService, rootDir, performWriteOperations);
+                processor.init(processorConfig);
+            }
+
+            logger.info("executing processors: ");
+            for (Processor processor : processors) {
+                logger.info("...." + processor.getClass().getName());
+                processor.process();
```
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/Main.java]: // yeah, two loops, could be done in one go, but KISS is life.
So we have clear phase of setting up everything and clear phase of doing something.

File: src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedIssuesEvaluator.java (hunk 8 above)
Comment (baranowb) [inline src/main/java/org/jboss/set/pull/processor/impl/evaluator/LinkedIssuesEvaluator.java]: https://github.com/jboss-set/pull-request-processor/issues/68

File: README.md [hunk 19]
```diff
@@ -7,12 +7,13 @@ It also checks the status of the latest merge on Hudson, post comments on github
...
 
-- -d: true or false (execution in dryRun mode for write operation in Github)
+- -w: true or false - determine if write pemission has been granted
 
-- -r: where to place the html report
+- -r: root path where processor can write output if it supports it
 
-java -jar -Daphrodite.config=${PULL_REQUEST_PROCESSOR_HOME}/aphrodite.json -Dstreams.json=${PULL_REQUEST_PROCESSOR_HOME}/streams.json /${PULL_REQUEST_PROCESSOR_HOME}/pull-processor.jar -s ${STREAMS} -r ${WORKSPACE}/report.html -d true  -as 7.0.z.GA
+Example: 
+java -jar -Daphrodite.config=${PULL_REQUEST_PROCESSOR_HOME}/aphrodite-test.json -Dstreams.json=${PULL_REQUEST_PROCESSOR_HOME}/streams.json ${PULL_REQUEST_PROCESSOR_HOME}/target/pull-processor-0.8.0-SNAPSHOT.jar -s jboss-eap-7.0.z[jbossas-jboss-eap7,jbossas-wildfly-core-eap] -p jboss-eap-7.0.z[jbossas-jboss-eap7,jbossas-wildfly-core-eap] -r ${PULL_REQUEST_PROCESSOR_HOME}/ -w false
```
[the remark this answers is not in the document]
Comment (baranowb) [reply README.md]: - -r: root path where processor can write output if it supports it  

Yeah, I know, but removing it and adding it back again make no sense.

[the remark this answers is not in the document]
Comment (soul2zimate) [reply README.md]: then it should keep ReportAction together with this option

**Comment by soul2zimate** [comment]
We can move it forward if there is no other change for this PR.

**Pull request merged**